src/MSBuildBinLogQuery/Parse/Parser.cs

```diff
@@ -0,0 +1,208 @@
+                return LogNodeType.Direct;
+            }
+            else if (_scanner.Token is DoubleSlashToken)
+            {
+                Consume<DoubleSlashToken>();
+                return LogNodeType.All;
+            }
+
+            throw new ParseException(_scanner.Expression);
+        }
+
+        private LogNode ParseLogNodeWithType(LogNodeType type)
+        {
+            if (_scanner.Token is MessageToken)
+            {
+                Consume<MessageToken>();
+                return new MessageNode(type);
+            }
+            else if (_scanner.Token is WarningToken)
+            {
+                Consume<WarningToken>();
+                return new WarningNode(type);
+            }
+            else if (_scanner.Token is ErrorToken)
+            {
+                Consume<ErrorToken>();
+                return new ErrorNode(type);
+            }
+
+            throw new ParseException(_scanner.Expression);
+        }
+
+        private LogNode ParseLogNode()
+        {
+            var type = ParseSlash();
+            return ParseLogNodeWithType(type);
+        }
+
+        private LogNode ParseLogNodeOrNull()
+        {
+            if (_scanner.Token is SingleSlashToken || _scanner.Token is DoubleSlashToken)
+            {
+                return ParseLogNode();
+            }
+
+            return null;
+        }
+
+        private AstNode ParseTaskNodeOrSingleSlashLogNode()
+        {
+            if (_scanner.Token is TaskToken)
+            {
+                Consume<TaskToken>();
+
+                var next = ParseLogNodeOrNull();
+                return new TaskNode(next);
+            }
+
+            return ParseLogNodeWithType(LogNodeType.Direct);
+        }
+
+        private AstNode ParseTaskNodeOrLogNodeOrNull()
+        {
+            if (_scanner.Token is SingleSlashToken)
+            {
+                Consume<SingleSlashToken>();
+                return ParseTaskNodeOrSingleSlashLogNode();
+            }
+            else if (_scanner.Token is DoubleSlashToken)
+            {
+                Consume<DoubleSlashToken>();
+                return ParseLogNodeWithType(LogNodeType.All);
+            }
+
+            return null;
+        }
+
+        private AstNode ParseTargetNodeOrTaskNodeOrSingleSlashLogNode()
+        {
+            if (_scanner.Token is TargetToken)
+            {
+                Consume<TargetToken>();
+
+                var next = ParseTaskNodeOrLogNodeOrNull();
+                return new TargetNode(next);
+            }
+            else if (_scanner.Token is TaskToken)
+            {
+                Consume<TaskToken>();
+
+                var next = ParseLogNodeOrNull();
+                return new TaskNode(next);
+            }
+
+            return ParseLogNodeWithType(LogNodeType.Direct);
+        }
+
+        private AstNode ParseTargetNodeOrTaskNodeOrLogNodeOrNull()
+        {
+            if (_scanner.Token is SingleSlashToken)
```

Comment:
personally I'd find pattern matching clearer here, but if you prefer these go for it

Reply:
What do you mean by "pattern matching" here (I think it has different meanings in different context..)? Could you give an example?

Reply:
I'm talking about [C# 7 `switch` type-based pattern matching](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/switch#type-pattern). This method, for example, could be expressed as:

```C#
        private AstNode ParseTargetNodeOrTaskNodeOrLogNodeOrNull()
        {
            switch (_scanner.Token)
            {
                case SingleSlashToken _:
                    Consume<SingleSlashToken>();
                    return ParseTargetNodeOrTaskNodeOrSingleSlashLogNode();
                case DoubleSlashToken _:
                    Consume<DoubleSlashToken>();
                    return ParseLogNodeWithType(LogNodeType.All);
            }

            return null;
        }
```

Reply:
I think It's better than what I have. Thanks!

Reply:
Done.